### PR TITLE
Allow followers to make HTTPS requests.

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
### Details
I think this works but I want to do a bunch of testing on Auth.

~**PREREQUISITE** see here, merge the two PRs mentioned: https://github.com/Expensify/Bedrock/pull/1641~

Migration strategy for this:
1. Merge this PR.
2. Merge: https://github.com/Expensify/Auth/pull/9815
3. Merge the bedrock PR to disable auto-escalating commands when they create HTTPS requests (basically a revert of 4b4d9966653138989a5e5e4b86e7a5573357f148)
4. Remove `NotLeading` exception from auth (this is just cleanup)
5. Remove `NotLeading` exception from bedrock (also just cleanup).

I think we need to resolve an issue with getting the `ClusterTester`s to work with one another first, as there are co-requisite changes in the two. Maybe break that out into a separate pair of PRs.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/350259

### Tests

_________
**Internal Testing Reminder:** when changing bedrock, please compile auth against your new changes

### CODE REVIEW

**Comment by flodnv** [comment]
>I think this works but I want to do a bunch of testing on Auth.

I assume this is done? 😄

File: BedrockCommand.cpp [hunk 1]
```diff
@@ -303,3 +306,59 @@ void BedrockCommand::setTimeout(uint64_t timeoutDurationMS) {
...
+        data["response"] = to_string(httpsRequest->response);
+        data["fullRequest"] = SEncodeBase64(httpsRequest->fullRequest.serialize());
+        data["fullResponse"] = SEncodeBase64(httpsRequest->fullResponse.serialize());
+        requests.push_back(SComposeJSONObject(data));
+    }
+
+    return SComposeJSONArray(requests);
+}
+
+string BedrockCommand::serializeData() const {
```
Comment (flodnv) [inline BedrockCommand.cpp:359]: I am confused why these 2 empty methods are here?

File: libstuff/SHTTPSManager.cpp [hunk 2]
```diff
@@ -133,16 +131,17 @@ void SStandaloneHTTPSManager::postPoll(fd_map& fdm, SStandaloneHTTPSManager::Tra
...
     created(STimeNow()),
     finished(0),
     timeoutAt(0),
     response(0),
     manager(manager_),
     sentTime(0),
-    requestID(SThreadLogPrefix)
+    requestID(requestID.empty() ? SThreadLogPrefix : requestID)
 {
+    // TODO: Remove this block to to enable HTTPS on followers. Also the `validate` method can be removed entirely.
```
Comment (flodnv) [inline libstuff/SHTTPSManager.cpp:144]: Will that be done as part of the attached issue?

File: sqlitecluster/SQLiteClusterMessenger.cpp [hunk 3]
```diff
@@ -137,6 +137,17 @@ bool SQLiteClusterMessenger::_sendCommandOnSocket(SHTTPSManager::Socket& socket,
 
     // This is what we need to send.
     SData request = command.request;
+
+    // If the command has https requests, we serialize them to escalate. In this case we also check if th command has data
```
Comment (flodnv) [inline sqlitecluster/SQLiteClusterMessenger.cpp:141]:
```suggestion
    // If the command has https requests, we serialize them to escalate. In this case we also check if the command has data
```

**Review by flodnv — COMMENTED** [review]
Seems strangely easy 😄

File: BedrockCommand.cpp (hunk 1 above)
Comment (tylerkaraszewski) [reply BedrockCommand.cpp:359]: The default command doesn't have anything to serialize. Commands that have things to serialize need to override these. Same idea as `peek` and `process` that do nothing here.

File: libstuff/SHTTPSManager.cpp (hunk 2 above)
Comment (tylerkaraszewski) [reply libstuff/SHTTPSManager.cpp:144]: This is step 3 in the description at the top of this PR.

**Comment by tylerkaraszewski** [comment]
> Seems strangely easy 😄

Doesn't it?

**Comment by tylerkaraszewski** [comment]
> > I think this works but I want to do a bunch of testing on Auth.
> 
> I assume this is done? 😄

It is all in the auth PR :)

**Review by flodnv — APPROVED** [review]
Tests are failing

**Comment by tylerkaraszewski** [comment]
Re-ran tests. Passed now.

File: BedrockServer.cpp [hunk 4]
```diff
@@ -2170,8 +2157,25 @@ unique_ptr<BedrockCommand> BedrockServer::buildCommandFromRequest(SData&& reques
...
+    request.erase("httpsRequests");
+    request.erase("serializedData");
+
     // Create a command.
     unique_ptr<BedrockCommand> command = getCommandFromPlugins(move(request));
+
+    // Apply HTTPS requests. If we had any, pretend we peeked this command in our current (likely LEADING) state.
+    if (serializedHTTPSRequests.size()) {
+        command->deserializeHTTPSRequests(serializedHTTPSRequests);
+        command->lastPeekedOrProcessedInState = _syncNode->getState();
```
Comment (rafecolton) [inline BedrockServer.cpp:2172]: I see elsewhere we set this variable we use `_server->getState()` - why do we use `_syncNode` here?

Comment (tylerkaraszewski) [reply BedrockServer.cpp:2172]: `_syncNode->getState();` is the underlying data for `_server->getState()` (see here):
https://github.com/Expensify/Bedrock/blob/66e0c3d16b20af54da8342440aaac56128849e3d/BedrockServer.cpp#L282-L283

and in `BedrockServer::getState`:
https://github.com/Expensify/Bedrock/blob/66e0c3d16b20af54da8342440aaac56128849e3d/BedrockServer.cpp#L2377-L2379

So the syncNode state is potentially "fresher". But `_syncNode` is private and not available from outside of `BedrockServer`. I think they will always be the same here anyway, so we could use either.

**Review by rafecolton — APPROVED** [review]
Did a little more Q&A with Tyler, this one is good to go! ✅

**Comment by tylerkaraszewski** [comment]
Going to merge, will prep the change to actually enable this after the auth change is merged and deployed.

**Pull request merged**